crates/proto/src/rr/domain/name.rs

```diff
@@ -899,20 +925,63 @@ impl LabelEnc for LabelEncUtf8 {
 }
 
 /// An iterator over labels in a name
-pub struct LabelIter<'a>(Iter<'a, Label>);
+pub struct LabelIter<'a> {
+    name: &'a Name,
+    index: usize,
+    started: bool,
+    finished: bool,
+}
 
 impl<'a> Iterator for LabelIter<'a> {
     type Item = &'a [u8];
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.0.next().map(Borrow::borrow)
+        if self.finished {
+            return None;
+        }
+        self.started = true;
+        let end = *self.name.label_ends.get(self.index)?;
+        let start = if self.index == 0 {
```

Comment:
Maybe consider writing this as a match statement matching self.index against 0 vs any other value?